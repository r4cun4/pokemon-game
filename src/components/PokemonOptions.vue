<template>
  <div class="options-container">

    <ul>
        <li
          v-for="pokemon in pokemons"
          :key="pokemon.id"
          @click="$emit( 'selectedPokemon', pokemon.id )"
          @mouseover="speechNames"
        >
          {{ pokemon.name }}
        </li>
    </ul>

<!-- colocar el :key es una buena practica para identificar el elemento de manera unica -->

  </div>
</template>

<script>
export default {
// Se definen las props que se pasaron como argumento en la PokemonPage en el llamado del componente
    props: {
        pokemons: {
            type: Array,
            required: true
        }
    },
    methods: {
        speechNames(e) {

            const msg = new SpeechSynthesisUtterance()
            msg.text = ''
            msg.lang = 'es-MX'

            if(e.target.matches('li:nth-child(1)')) {
                msg.text = this.pokemons[0].name;
                console.log(msg)
                window.speechSynthesis.speak(msg);
            } else if(e.target.matches('li:nth-child(2)')) {
                msg.text = this.pokemons[1].name;
                console.log(msg)
                window.speechSynthesis.speak(msg);
            } else if(e.target.matches('li:nth-child(3)')) {
                msg.text = this.pokemons[2].name;
                console.log(msg)
                window.speechSynthesis.speak(msg);
            } else {
                msg.text = this.pokemons[3].name;
                console.log(msg)
                window.speechSynthesis.speak(msg);
            }
        }
    }
}
</script>

<style scoped>

    ul {
        display: flex;
        flex-direction: column;
        gap: 15px;
        list-style-type: none;
        padding: 0;
    }
    li {
        background-color: #fff;
        border-radius: inherit;
        border: 2px solid red;
        cursor: pointer;
        padding: 10px;
        width: 250px;
        text-transform: capitalize;
        font-weight: bold;
        color: gray;
    }

    li:hover {
        background-color: red;
        border: 2px solid red;
        color: #fff;
        transition: 1s;
    }

    .options-container {
        display: flex;
        justify-content: center;
    }

</style>