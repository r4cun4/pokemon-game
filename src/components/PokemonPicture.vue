<template>
    <div class="pokemon-container">

        <img :src="imgSrc"
            class="hidden-pokemon"
            alt="pokemon">

        <img
          v-if="showPokemon"
          :src="imgSrc"
          class="fade-in"
          alt="pokemon">

    </div>

    
</template>

<script>
export default {

// en las props recibo el argumento pokemonID que lo estamos pasando en la pokemonPage en el llamado del componente PokemonPicture

// en este caso lo definimos como Number ya que el endpoint que se usa en cada call a la API termina en un numero

// la property showPokemon es para poder mostrar la imagen dependiendo es booleana ya que si es true mostrara la imagen y si es false no

    props: {
        pokemonId: {
            type: Number,
            required: true
        },
        showPokemon: {
            type: Boolean,
            required: true,
            default: false
        }
    },
// pasamos la property con interpolacion para que sea dinamica
    computed: {
        imgSrc() {
            return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/${this.pokemonId}.svg`
        }
    }

}
</script>

<style scoped>

    .pokemon-container {
        display: flex;
        position: relative;
        justify-content: center;
        height: 200px;
    }
    img {
        height: 200px;
        position: absolute;
        user-select: none;
        -moz-user-select: none;
        -ms-user-select: none;
        -webkit-user-drag: none;
        -webkit-user-select: none;
    }
    .hidden-pokemon {
        filter: brightness(0);
    }

</style>