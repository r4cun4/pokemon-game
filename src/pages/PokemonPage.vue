<template>

    <h1 v-if="!pokemon">Espere por favor...</h1>

    <div v-else>
        <h1>¿Quién es este Pokémon?</h1>
        <!-- los atributos que se le pasan al componente deben escribirse en kebab-case -->
        <PokemonPicture
          :pokemon-id="pokemon.id"
          :show-pokemon="showPokemon"
        />

        <PokemonOptions
          :pokemons="pokemonArr"
          @selected-pokemon="checkAnswer"
          @click="speechAnswer"
        />

        <div v-if="showAnswer">

            <h2 class="fade-in">{{ message }}</h2>
            <button
              @click="newGame">Nuevo Juego
            </button>

        </div>

    </div>

</template>

<script>
import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'

import getPokemonOptions from '@/helpers/getPokemonOptions'

console.log( getPokemonOptions() )

export default {
    components: { PokemonPicture, PokemonOptions },
    data() {
        return {
            pokemonArr: [],
            // pokemon tiene el id y el nombre, se inicializa null porque no sabemos q pokemon sera el de la img
            pokemon: null, 
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },
    methods: {
        async mixPokemonArray() {
            this.pokemonArr = await getPokemonOptions()
            
            const rndInt = Math.floor( Math.random() * 4 )
            this.pokemon = this.pokemonArr[ rndInt ]
        },
        checkAnswer( selectedId ) {
            this.showAnswer = true

            if( selectedId === this.pokemon.id ) {
                this.showPokemon = true
                this.message = `Muy bien Lolita! es, ${ this.pokemon.name[0].toUpperCase() }` + `${this.pokemon.name.substring(1)}`
            } else {
                this.message = `Ese no era Loli!, era ${ this.pokemon.name.toUpperCase() }`
            }
        },
        speechAnswer() {
            const msg = new SpeechSynthesisUtterance();
            msg.text = this.message;
            window.speechSynthesis.speak(msg);
        },
        newGame() {
            this.mixPokemonArray(),
            this.showPokemon = false,
            this.showAnswer = false,
            this.message = ''
            this.pokemonArr = [],
            this.pokemon = null
        }
    },
    // El metodo mounted ejecuta el codigo que tenga una vez que el componente este montado
    mounted() {
        this.mixPokemonArray()
    }
}
</script>

<style scoped>

button {
    width: 60px;
    height: 60px;
    background-image: url('@/assets/restart-icon.svg');
    background-repeat: no-repeat;
    background-size: contain;
    background-position: center;
    border: none;
    color: transparent;
}


</style>