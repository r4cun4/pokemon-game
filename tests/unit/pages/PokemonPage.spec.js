import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from "@/pages/PokemonPage"
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'

import { mockPokemons, pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage Component', () => {

    let wrapper;
    
    beforeEach(() => {
        wrapper = shallowMount( PokemonPage )
    })

    test("debe de hacer match con el snapshot", () => {

        expect(wrapper.html).toMatchSnapshot();
    
    });

    test("debe de llamar mixPokemonArray al montar", () => {

        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray')
        // se vuelve a ejecutar el shallowMount para que primero pueda espiar el metodo mixPokemonArray antes de que se monte la pagina
        shallowMount( PokemonPage )

        expect( mixPokemonArraySpy ).toHaveBeenCalled()
    
    });

    test("debe de hacer match con el snapshot cuando cargan los pokemons", () => {

        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons,
                    pokemon: mockPokemons[0], 
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            },
        })

        expect( wrapper.html() ).toMatchSnapshot()
    
    });

    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {

        //Tarea 111

        //PokemonPicture debe existir
        //PokemonOptions debe existir

        //PokemonPicture attribute pokemon === 5
        //PokemonOptions attribute pokemons toBe true

        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons,
                    pokemon: mockPokemons[0], 
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            },
        })

        // Busco los componentes
        const pictureComponent = wrapper.findComponent(PokemonPicture)
        const optionsComponent = wrapper.findComponent(PokemonOptions)

        console.log('Soy el componente PokemonPicture', pictureComponent.html())
        console.log('Soy el componente PokemonOptions', optionsComponent.html())

        // desestructuro el id
        const { id } = mockPokemons[0]

        console.log('soy el id', id)

        expect(pictureComponent.exists()).toBeTruthy()
        expect(optionsComponent.exists()).toBeTruthy()
        expect( id ).toBe(5)
        expect(optionsComponent.attributes('pokemons')).toBeTruthy()

        // Solucion FH

        // const pictureComponent = wrapper.find('pokemon-picture-stub')
        // const optionsComponent = wrapper.find('pokemon-options-stub')

        // expect(pictureComponent.exists()).toBeTruthy()
        // expect(optionsComponent.exists()).toBeTruthy()

        // expect( pictureComponent.attributes('pokemonid') ).toBe('5')
        // expect( optionsComponent.attributes('pokemons') ).toBeTruthy()

    })

    test("pruebas con checkAnswer", async() => {

        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons,
                    pokemon: mockPokemons[0], 
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            },
        })

        await wrapper.vm.checkAnswer(5)
        console.log(wrapper.vm)

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`Muy bien Lolita! es, ${ mockPokemons[0].name }`)


        // tarea 112 - evaluar la salida de checkAnswer con otro pokemon

        await wrapper.vm.checkAnswer(10)

        // vm permite evaluar props reactivas de la funcion data
        expect(wrapper.vm.message).toBe(`Ese no era Loli!, era ${mockPokemons[0].name.toUpperCase()}`)
    
    });

})