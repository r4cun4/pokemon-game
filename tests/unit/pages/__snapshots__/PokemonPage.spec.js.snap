// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe de hacer match con el snapshot 1`] = `[Function]`;

exports[`PokemonPage Component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>¿Quién es este Pokémon?</h1><!-- los atributos que se le pasan al componente deben escribirse en kebab-case -->
  <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
